examples/trusted-gateway-token/contracts/contracts/Token.sol

```diff
@@ -0,0 +1,122 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.4;
+
+import "hardhat/console.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "hardhat/console.sol";
+/**
+ * @title Token
+ * @dev Very simple ERC20 Token example, where all tokens are pre-assigned to the creator.
+ * Note they can later distribute these tokens as they wish using `transfer` and other
+ * `ERC20` functions.
+ * Based on https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v2.5.1/contracts/examples/SimpleToken.sol
+ */
+contract Token is ERC20, Ownable {
+  string public url;
+  address private _signer;
+  mapping(address=>bool) claimed;
+  error OffchainLookup(string url, bytes prefix);
+  struct BalanceProof {
+    bytes signature;
+    uint balance;
+  }
+
+  /**
+    * @dev Constructor that gives msg.sender all of existing tokens.
+    */
+  constructor(
+      string memory name,
+      string memory symbol,
+      uint256 initialSupply
+  ) ERC20(name, symbol) {
+    _mint(msg.sender, initialSupply);
+  }
+
+  function setUrl(string memory url_) external onlyOwner{
+    url = url_;
+  }
+
+  function setSigner(address signer_) external onlyOwner{
+    _signer = signer_;
+  }
+
+  function getSigner() external view returns(address){
+    return _signer;
+  }
+
+  function balanceOf(address addr) public override view returns(uint balance) {
+    if(claimed[addr]){
+      return super.balanceOf(addr);
+    }else{
+      revert OffchainLookup(url,
+        // Selector does not seem to work against public function
```

Comment:
Why doesn't it work here, but it does work in `transfer`?

Reply:
`transferWithProof` is `external` while `balanceOfWithProof` is `public` so that `transferWithProof` can call the function.